### PR TITLE
Added authentication to Allow list call

### PR DESCRIPTION
**Pull request by bharney0**
Added authentication to Allow list call
[AB#225965](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/225965)
[AB#223857](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/223857)

### CODE REVIEW

File: docs-markdown/src/helper/AllowList.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		if (!allowlist) {
+			// if user is inside :::image::: tag, then ask them for quickpick of products based on allow list
+			// call allowlist with API Auth Token
+			// you will need auth token to call list
+			const token: TokenResponse = this.context.globalState.get('token');
+			const config = {
+				headers: { Authorization: `Bearer ${token.accessToken}` }
+			};
+			const response = await Axios.get(
+				'https://docs.microsoft.com/api/metadata/allowlists',
```
Comment (OsmondJiang) [inline docs-markdown/src/helper/AllowList.ts:22]: https://docsvalidation.azurefd.net/validation/allowlists

File: docs-markdown/src/helper/Auth.ts [hunk 2]
```diff
@@ -40,9 +35,9 @@ export class Auth {
 			const authContext = new AuthenticationContext(this.authorityUrl);
 			return new Promise((resolve, reject) => {
 				authContext.acquireTokenWithClientCredentials(
```
Comment (OsmondJiang) [inline docs-markdown/src/helper/Auth.ts:37]: our resource id is: 79abd399-c39f-4589-8dda-ce6f7d6f8d70

**Comment by OsmondJiang** [comment]
please verify the JWT token you got via https://jwt.io and check if you get the `access_as_application` role.

**Comment by bharney0** [comment]
> please verify the JWT token you got via https://jwt.io and check if you get the `access_as_application` role.

@OsmondJiang I believe everything is setup correctly. I see your permissions to the Docs Validation Management in our App on portal side. And our payload from the token looks like this: 
image Which seems to be the app.

File: docs-markdown/src/helper/Auth.ts [hunk 3]
```diff
@@ -16,36 +17,35 @@ export interface TokenResponse {
...
-		if (!token || token.expiresIn < expiresTime) {
+		const expiresTime = Math.round(new Date().getTime() / 1000);
+		if (
+			!token ||
+			(token.expiresIn &&
+				Math.round((new Date().getTime() + token.expiresIn * 1000) / 1000) < expiresTime)
+		) {
 			const authContext = new AuthenticationContext(this.authorityUrl);
 			return new Promise((resolve, reject) => {
 				authContext.acquireTokenWithClientCredentials(
```
Comment (OsmondJiang) [inline docs-markdown/src/helper/Auth.ts:42]: Here you should get the access token of our AAD application, not the Microsoft Graph, so you need change the **resource** to our application id url:

Sandbox for testing: api://18c498b3-2ced-4ea0-b6ea-c52841b603de
PROD: api://79abd399-c39f-4589-8dda-ce6f7d6f8d70

I referenced from this page: https://docs.microsoft.com/en-us/javascript/api/overview/azure/activedirectory?view=azure-node-latest

image

Comment (OsmondJiang) [reply docs-markdown/src/helper/Auth.ts:42]: Example code: https://github.com/AzureAD/azure-activedirectory-library-for-nodejs/blob/master/sample/client-credentials-sample.js

Comment (bharney0) [reply docs-markdown/src/helper/Auth.ts:42]: @OsmondJiang Thank you for that. Once I switched the resouce to the api:// url the authorization to the allowlist started working.

Comment (OsmondJiang) [reply docs-markdown/src/helper/Auth.ts:42]: 👍👍 great to see that, if you verified the Sandbox is working, then PROD should be OK.

**Pull request merged**